include/bmi/Bmi_Py_Adapter.hpp

```diff
@@ -44,6 +47,10 @@ namespace models {
             Bmi_Py_Adapter(Bmi_Py_Adapter const&) = delete;
             Bmi_Py_Adapter(Bmi_Py_Adapter&&) = delete;
 
+            ~Bmi_Py_Adapter() override {
+                Finalize();
+            }
```

Comment:
Unlike Bmi_Adapter, it should be fine to call `Finalize()` here, since we are calling the "final" version here. To make it safer, `Bmi_*_Adapter` classes should probably have the `final` specifier in the declaration.

Reply:
I'd like to see those `final` specifiers actually applied.